### PR TITLE
Tune debug info configuration to work for package releases on Linux.

### PR DESCRIPTION
**Pull request by stellaraccident**
This does several things:

* Adds a CMake file to do some global system compiler setup and put the `CMAKE_MSVC_DEBUG_INFORMATION_FORMAT` from the README in there for Windows (this file was doing more and I've started this half a dozen times but never finished, so just going to land it now).
* Ensures that `-Wl,--build-id` is set on any binaries that we are doing debug symbol splitting from.
* Hardcodes CMP0087 (enabling generator expressions in install CODE) to NEW. Some of our very old projects need this and it unfortunately is a policy that must be set globally vs scoped.
* Adds a `{project_name}_BUILD_TYPE` override, allowing fine grained setting of build type per project. Used in our packaging preset to have the compiler build in Release vs RelWithDebInfo (which saves us some 600MB of debug info for the common case where we trust the compiler or debug it with reproducers vs stack traces).
* Adds a `THEROCK_QUIET_INSTALL` defaulted to ON (present behavior) and sets it to OFF in our packaging preset as important diagnostic information is displayed at install time for packaging cases.
* Fixes up some toochain init flags and sets DEBUG/RELWITHDEBINFO flag defaults for a new `THEROCK_MINIMAL_DEBUG_INFO` configure option (enabling `-g1` debug info). Note that there may be more tuning needed here for precise incantation, but it gets very compiler specific from here when tuning debug flags.
* Adds a preset file where we can put official combinations used for packaging builds, etc.

A release packaging build can now be done with:

```
cmake --preset linux-release-package -Bbuild -DTHEROCK_AMDGPU_FAMILIES=gfx1100 -DCMAKE_C_COMPILER_LAUNCHER=ccache  -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
```

More tuning is possible, but this gets us debug symbols that should be useful and are not of a size that is a non-starter:

```
$ du -h -d 0 *_dbg_*
54M     amd-llvm_dbg_generic
6.3M    base_dbg_generic
464M    blas_dbg_gfx1100
4.8M    core-hip_dbg_generic
3.5M    core-runtime_dbg_generic
25M     fft_dbg_gfx1100
9.8M    hipify_dbg_generic
928K    host-blas_dbg_generic
152K    host-suite-sparse_dbg_generic
34M     miopen_dbg_gfx1100
840M    prim_dbg_gfx1100
5.0M    rand_dbg_gfx1100
2.2M    rccl_dbg_gfx1100
40M     rocprofiler-sdk_dbg_generic
```

Predictably, the C++ template heavy prim is very bad. May add more project specific tuning, compiler specific flag setup to enable `-gmlt` with less debug info, or handle test/client debug symbols differently.

### CODE REVIEW

File: cmake/therock_compiler_config.cmake [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Project wide compiler configuration.
+
+# On win32 only support embedded debug databases project wide.
+if(WIN32 AND NOT DEFINED CMAKE_MSVC_DEBUG_INFORMATION_FORMAT)
+  set(CMAKE_MSVC_DEBUG_INFORMATION_FORMAT "$<$<CONFIG:Debug,RelWithDebInfo>:Embedded>")
+endif()
```
Comment (ScottTodd) [inline cmake/therock_compiler_config.cmake]: I've only built in `Release` on Windows and I've still needed this flag to get ccache hits. Maybe one of the subprojects configured itself to include debug information.

We could apply the option using a mechanism like https://github.com/iree-org/iree/blob/66f9750fad2a7ed7d5ea2ba1bdd255ed319156a5/build_tools/cmake/iree_copts.cmake#L553-L570 , also forcing the option on if

```cmake
# Maybe check for sccache also?
if(MSVC)
  if("${CMAKE_C_COMPILER_LAUNCHER}" MATCHES "ccache" OR
     "${CMAKE_CXX_COMPILER_LAUNCHER}" MATCHES "ccache")
```

(The direct manipulation of compiler flags in that IREE code should be replaced with `CMAKE_MSVC_DEBUG_INFORMATION_FORMAT` now too)

See also https://github.com/mozilla/sccache and search "The process for using sccache with MSVC and cmake".

File: CMakePresets.json
Comment (ScottTodd) [inline CMakePresets.json:1]: I'll pull this down and test locally a bit. VSCode defaulted to discovering and using CMake presets when last I checked, resulting in unexpected/annoying behavior...

**Review by ScottTodd — COMMENTED** [review]
Nice changes. I'll test the noted edge cases a bit myself.

File: CMakePresets.json
Comment (ScottTodd) [reply CMakePresets.json:1]: Docs: https://github.com/microsoft/vscode-cmake-tools/blob/main/docs/cmake-presets.md

Logs after checking out this branch (`cmake.userCMakePresets` value of `auto`)
```
[presetController] Reading and validating the presets "file D:\projects\TheRock\CMakePresets.json"
[presetController] Successfully validated D:\projects\TheRock\CMakePresets.json against presets schema
[presetController] Expanding presets file D:\projects\TheRock\CMakePresets.json
[presetController] Successfully expanded presets file D:\projects\TheRock\CMakePresets.json
```

That was fine... until I rebooted VSCode:
image

Now I can't configure/build without selecting a preset.

Ideas:

* I can just set `cmake.userCMakePresets` to `never`
* We could add a null/base preset

Comment (ScottTodd) [reply CMakePresets.json:1]: Ugh. 

```json
    {
      "name": "base",
      "displayName": "Base CMake settings",
      "generator": "Ninja"
    },
```

When CMake presets are enabled, the vscode-cmake-tools extension disables variants and kits, so I can't select between Debug/RelWithDebInfo/Release _or_ select a compiler...
```
[cmake] -- The C compiler identification is GNU 13.2.0
[cmake] -- The CXX compiler identification is GNU 13.2.0
```

Comment (ScottTodd) [reply CMakePresets.json:1]: Trying something like this now:
```json
    {
      "name": "windows-release",
      "displayName": "Settings for Windows release builds",
      "generator": "Ninja",
      "cacheVariables": {
        "CMAKE_BUILD_TYPE": "Release",
        "CMAKE_C_COMPILER": "cl.exe",
        "CMAKE_CXX_COMPILER": "cl.exe"
      },
      "condition":{
          "type": "equals",
          "lhs": "${hostSystemName}",
          "rhs": "Windows"
      }
    }
```

Seems to be doing something?
```
[preset] Configure preset windows-release: No architecture specified for cl.exe, using x64 by default
[preset] Configure preset windows-release: No toolset architecture specified for cl.exe, using "host=x64" by default
[preset] Using developer environment from Visual Studio (instance 40d03a2a, version 17.12.35707.178, installed at "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools")
[preset] Configure preset windows-release: No architecture specified for cl.exe, using x64 by default
```

```
[driver] NOTE: You are configuring with preset Settings for Windows release builds, but there are some overrides being applied from your VS Code settings.
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER=cl.exe -DCMAKE_CXX_COMPILER=cl.exe -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -UCMAKE_MSVC_DEBUG_INFORMATION_FORMAT -DPython3_EXECUTABLE=d:/projects/TheRock/.venv/Scripts/python -UCMAKE_LINKER_TYPE -DTHEROCK_VERBOSE=ON -DTHEROCK_RESET_FEATURES=ON -DBUILD_TESTING=ON -DTHEROCK_ENABLE_COMM_LIBS=OFF -DTHEROCK_ENABLE_ML_LIBS=ON -DTHEROCK_AMDGPU_FAMILIES=gfx110X-dgpu -UCMAKE_PREFIX_PATH -SD:/projects/TheRock -BD:/projects/TheRock/out/build/windows-release -G Ninja
[cmake] -- The C compiler identification is MSVC 19.42.34436.0
[cmake] -- The CXX compiler identification is MSVC 19.42.34436.0
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - done
[cmake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2022/BuildTools/VC/Tools/MSVC/14.42.34433/bin/Hostx64/x64/cl.exe - skipped
```

Comment (ScottTodd) [reply CMakePresets.json:1]: Alright, I have something workable:

```json
    {
      "name": "windows-base",
      "displayName": "Windows base preset",
      "description": "Ninja generator, MSVC (cl.exe), x64 architecture",
      "generator": "Ninja",
      "binaryDir": "${sourceDir}/build",
      "architecture": {
        "value": "host=x64",
        "strategy": "external"
      },
      "toolset": {
        "value": "host=x64",
        "strategy": "external"
      },
      "cacheVariables": {
        "CMAKE_C_COMPILER": "cl.exe",
        "CMAKE_CXX_COMPILER": "cl.exe"
      },
      "condition":{
          "type": "equals",
          "lhs": "${hostSystemName}",
          "rhs": "Windows"
      }
    },
    {
      "name": "windows-release",
      "displayName": "Windows release builds preset",
      "description": "Ninja generator, MSVC (cl.exe), x64 architecture, Release",
      "cacheVariables": {
        "CMAKE_BUILD_TYPE": "Release"
      },
      "inherits": [
        "windows-base"
      ]
    }
```

image


Looking into ccache and the `CMAKE_MSVC_DEBUG_INFORMATION_FORMAT` setting next.

Comment (ScottTodd) [reply CMakePresets.json:1]: Ugh. Presets default the build/binary directory to `TheRock\out\build\{PRESET_NAME}`

File: cmake/therock_compiler_config.cmake (hunk 1 above)
Comment (ScottTodd) [reply cmake/therock_compiler_config.cmake]: Okay, I think this way of setting the CMake flag for ccache is actually working now, at least for me locally. We can watch the CI to make sure it still gets cache hits too.

File: cmake/therock_compiler_config.cmake [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Project wide compiler configuration.
+
+# On win32 only support embedded debug databases project wide.
```
Comment (ScottTodd) [inline cmake/therock_compiler_config.cmake:3]: suggested context, then we can trim the docs
```suggestion
# On win32 only support embedded debug databases project wide.
# This improves compatibility with ccache and sccache:
# https://github.com/ccache/ccache/issues/1040
```

https://github.com/ROCm/TheRock/blob/334f6c36eb1731272a310fb944759c02c6361f33/docs/development/windows_support.md?plain=1#L207-L224

File: CMakePresets.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        "amd-llvm_BUILD_TYPE": "Release",
+        "therock-host-blas_BUILD_TYPE": "Release",
+        "therock-SuiteSparse_BUILD_TYPE": "Release",
+        "THEROCK_SPLIT_DEBUG_INFO": "ON",
+        "THEROCK_MINIMAL_DEBUG_INFO": "ON",
+        "THEROCK_QUIET_INSTALL": "OFF"
+      }
+    }
+  ]
+}
```
Comment (ScottTodd) [inline CMakePresets.json:24]: Here are my suggested edits to get this working for my VSCode setup on Windows

```suggestion
  "configurePresets": [
    {
      "name": "linux-release-package",
      "displayName": "Linux release packages",
      "description": "Ninja generator, default compiler, RelWithDebInfo for most subprojects",
      "generator": "Ninja",
      "cacheVariables": {
        "CMAKE_BUILD_TYPE": "RelWithDebInfo",
        "amd-llvm_BUILD_TYPE": "Release",
        "therock-host-blas_BUILD_TYPE": "Release",
        "therock-SuiteSparse_BUILD_TYPE": "Release",
        "THEROCK_SPLIT_DEBUG_INFO": "ON",
        "THEROCK_MINIMAL_DEBUG_INFO": "ON",
        "THEROCK_QUIET_INSTALL": "OFF"
      }
    },
    {
      "name": "windows-base",
      "displayName": "Windows base preset",
      "description": "Ninja generator, MSVC (cl.exe), x64 architecture",
      "generator": "Ninja",
      "binaryDir": "${sourceDir}/build",
      "architecture": {
        "value": "host=x64",
        "strategy": "external"
      },
      "toolset": {
        "value": "host=x64",
        "strategy": "external"
      },
      "cacheVariables": {
        "CMAKE_C_COMPILER": "cl.exe",
        "CMAKE_CXX_COMPILER": "cl.exe"
      },
      "condition": {
        "type": "equals",
        "lhs": "${hostSystemName}",
        "rhs": "Windows"
      }
    },
    {
      "name": "windows-release",
      "displayName": "Windows release builds preset",
      "description": "Ninja generator, MSVC (cl.exe), x64 architecture, Release",
      "cacheVariables": {
        "CMAKE_BUILD_TYPE": "Release"
      },
      "inherits": [
        "windows-base"
      ]
    }
  ]
}

```

**Review by ScottTodd — COMMENTED** [review]
LGTM if my suggestions are applied

File: CMakePresets.json (hunk 3 above)
Comment (stellaraccident) [reply CMakePresets.json:24]: Let's follow up with these changes. There is something odd about the formatting in the suggestion.

Comment (ScottTodd) [reply CMakePresets.json:24]: Sent https://github.com/ROCm/TheRock/pull/718
